// react-native-woodpicker/dist/components/Picker.js
import React, { useState, useRef, useMemo } from "react";
import { Animated } from "react-native";
import { Picker as RNPicker } from "@react-native-picker/picker";
import DefaultInputButton from "./InputButton";
import DefaultDoneBar from "./DoneBar";
import IOSPicker from "./IOSPicker";
import AndroidPicker from "./AndroidPicker";
import { DEFAULT_BACKDROP_ANIMATION } from "../constants/animation";
import { getAnimatedProperties } from "../helpers/animation";
import { isIOS } from "../helpers/iphone";
import { EMPTY_ITEM } from "../constants/item";
const Picker = ({ item, items, onItemChange, disabled = false, placeholder = "", title = "", mode = "dialog", doneButtonLabel = "", isNullable = false, onOpen = () => null, onClose = () => null, style, containerStyle, textInputStyle, touchableStyle, InputComponent, DoneBarComponent, backdropAnimation = DEFAULT_BACKDROP_ANIMATION, androidCustomProps, iosCustomProps, }) => {
    const [show, setShow] = useState(false);
    const [selectedItem, setSelectedItem] = useState(item
        ? items.find(({ label, value }) => label === item.label && value === item.value) || EMPTY_ITEM
        : isNullable
            ? EMPTY_ITEM
            : items[0]);
    const fadeAnimationValue = useRef(new Animated.Value(0)).current;
    const animationProperties = useMemo(() => getAnimatedProperties(backdropAnimation), [backdropAnimation]);
    const handleItemChange = (value, index) => {
        const nullableItems = isNullable ? [EMPTY_ITEM, ...items] : items;
        const newSelectedItem = nullableItems.find((currentItem) => value === currentItem.value) ||
            EMPTY_ITEM;
        if (!isIOS) {
            onItemChange(newSelectedItem, index);
        }
        setSelectedItem(newSelectedItem);
    };
    const toggle = () => {
        setShow(!show);
        show ? onOpen === null || onOpen === void 0 ? void 0 : onOpen() : onClose === null || onClose === void 0 ? void 0 : onClose();
    };
    const togglePicker = () => {
        if (disabled) {
            return;
        }
        if (!show) {
            toggle();
        }
        Animated.timing(fadeAnimationValue, {
            toValue: !show ? animationProperties.opactiy : 0,
            duration: !show ? animationProperties.duration : 0,
            delay: !show ? animationProperties.delay : 0,
            useNativeDriver: true,
        }).start(show ? toggle : undefined);
    };
    const onDonePress = () => {
        togglePicker();
        const itemIndex = (isNullable ? [EMPTY_ITEM, ...items] : items).findIndex(({ value }) => value === selectedItem.value);
        onItemChange(selectedItem, itemIndex);
    };
    const renderPickerItems = () => {
        const tempItems = isNullable ? [EMPTY_ITEM, ...items] : items;
        return tempItems.map((current) => {
            return (React.createElement(RNPicker.Item, { label: current.label, value: current.value, key: current.label }));
        });
    };
    const renderPlaceholder = () => {
        if (item && item.label) {
            return item.label;
        }
        return placeholder;
    };
    const renderInputButton = () => {
        const inputProps = {
            togglePicker,
            style,
            text: renderPlaceholder(),
            textInputStyle,
            touchableStyle,
            isNullable: false,
            resetValue: undefined,
            renderHiddenCompactIOSPicker: () => null, //DatePicker only
        };
        const RenderComponent = InputComponent
            ? InputComponent
            : DefaultInputButton;
        return React.createElement(RenderComponent, Object.assign({}, inputProps));
    };
    const renderDoneBarButton = () => {
        const barProps = {
            title,
            doneButtonLabel,
            onDonePress,
        };
        const RenderComponent = DoneBarComponent
            ? DoneBarComponent
            : DefaultDoneBar;
        return React.createElement(RenderComponent, Object.assign({}, barProps));
    };
    const pickerProps = {
        selectedItem,
        disabled,
        title,
        renderPickerItems,
        renderInput: renderInputButton,
        onItemChange: handleItemChange,
        togglePicker,
        containerStyle,
    };
    if (isIOS) {
        const iosProps = Object.assign(Object.assign({}, pickerProps), { show, renderDoneBar: renderDoneBarButton, togglePicker,
            containerStyle, animationValue: fadeAnimationValue, customProps: iosCustomProps || {} });
        return React.createElement(IOSPicker, Object.assign({}, iosProps));
    }
    const androidProps = Object.assign(Object.assign({}, pickerProps), { mode, customProps: androidCustomProps || {} });
    return React.createElement(AndroidPicker, Object.assign({}, androidProps));
};
export default Picker;
